## Data Cleaning

### **Import libraries**

In [29]:
!pip install pymongo

import pandas as pd
from pymongo import MongoClient

### **Reading Tables**

In [30]:
from google.colab import drive
drive.mount('/content/drive')

data_2020 = pd.read_csv('/content/drive/MyDrive/2020EAVSData.csv')
data_2016 = pd.read_csv('/content/drive/MyDrive/2016EAVSData.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


<ipython-input-30-887f4565c29a>:5: DtypeWarning:

Columns (4,7,8,9,11,12,14,15,16,17,18,21,22,24,25,26,27,29,31,34,35,36,42,44,46,48,49,51,53,65,67,69,71,73,75,87,89,91,93,95,97,109,111,113,115,117,119,127,129,131,141,143,145,147,150,151,152,153,155,157,158,159,160,161,163,164,165,166,168,170,173,177,181,185,187,188,189,190,191,192,194,195,196,198,199,200,204,206,207,208,209,210,213,214,216,219,220,221,223,224,225,227,228,229,233,235,236,237,238,239,240,242,243,244,245,246,250,251,252,253,255,258,259,266,268,269,271,274,276,278,280,281,283,285,287,289,290,291,292,293,299,302,303,304,305,306,307,308,309,313,314,316,317,318,319,325,332,333,335,337,339,341,343,345,346,347,348,349,350,351,353,354,355,357,359,361,362,364,368,373,375,377,378,379,380,382,383,386,387,388,390,391,392,394,395,396,398,400,401,402,404,406,408,409,410,414,416,418,421,423,425,428,429,430,432,433,434,436,437,438,440,441,442,444,447,449,452,453,454,456,457,458,460,461,462,464,465,466,468,469,474,475,476,477,479,480,48

### **Renaming columns and dropping unnecessary columns**

In [31]:
# Defining the columns to keep
columns_2020 = [
    'FIPSCode', 'Jurisdiction_Name', 'State_Full', 'State_Abbr',
    'C1a', 'C1b', 'C1c', 'C1d', 'C1e', 'C1f', 'C2a', 'C3a',
    'C4a', 'C4b', 'C4c', 'C4d', 'C4e', 'C4f', 'C4g', 'C4h', 'C4i', 'C4j', 'C4k',
    'C4l', 'C4m', 'C4n', 'C4o'
]

columns_to_rename_2020 = {
    'C1a': 'By-mail Transmitted Total',
    'C1b': 'By-mail Returned By Voters Total',
    'C1c': 'By-mail Returned Undeliverable',
    'C1d': 'By-mail Voided',
    'C1e': 'By-mail Voted In-Person',
    'C1f': 'By-mail Status Unknown',
    'C2a': 'Total Transmitted Permanent By-mail Registrants',
    'C3a': 'By-mail Counted Total',
    'C4a':	'By-mail Rejected Total',
    'C4b':	'By-mail Rejected Deadline',
    'C4c':	'By-mail Rejected Voter Signature',
    'C4d':	'By-mail Rejected Witness Signature',
    'C4e':	'By-mail Rejected Non-matching Signature',
    'C4f':	'By-mail Rejected No EO Signature',
    'C4g':	'By-mail Rejected Unofficial Envelope',
    'C4h':	'By-mail Rejected Ballot Missing',
    'C4i':	'By-mail Rejected Envelope Not Sealed',
    'C4j':	'By-mail Rejected No Address',
    'C4k':	'By-mail Rejected Multiple Ballots',
    'C4l':	'By-mail Rejected Deceased',
    'C4m':	'By-mail Rejected Already Voted',
    'C4n':	'By-mail Rejected No Voter ID',
    'C4o':	'By-mail Rejected No Ballot App'
}

columns_2016 = [
    'FIPSCode', 'FIPS_2Digit', 'State', 'JurisdictionName',
    'C1a', 'C1b', 'C1c', 'C1d', 'C1e', 'C3', 'C4a', 'C4b',
    'C5a', 'C5b', 'C5c', 'C5d', 'C5e', 'C5f', 'C5g', 'C5h', 'C5i', 'C5j', 'C5k', 'C5l', 'C5m', 'C5n'
]

columns_to_rename_2016 = {
    'C1a': 'By-mail Transmitted Total',
    'C1b': 'By-mail Returned By Voters Total',
    'C1c': 'By-mail Returned Undeliverable',
    'C1d': 'By-mail Voided',
    'C1e': 'By-mail Status Unknown',
    'C3': 'Total Transmitted Permanent By-mail Registrants',
    'C4a': 'By-mail Counted Total',
    'C4b':	'By-mail Rejected Total',
    'C5a':	'By-mail Rejected Deadline',
    'C5b':	'By-mail Rejected Voter Signature',
    'C5c':	'By-mail Rejected Witness Signature',
    'C5d':	'By-mail Rejected Non-matching Signature',
    'C5e':	'By-mail Rejected No EO Signature',
    'C5f':	'By-mail Rejected Unofficial Envelope',
    'C5g':	'By-mail Rejected Ballot Missing',
    'C5h':	'By-mail Rejected Envelope Not Sealed',
    'C5i':	'By-mail Rejected No Address',
    'C5j':	'By-mail Rejected Multiple Ballots',
    'C5k':	'By-mail Rejected Deceased',
    'C5l':	'By-mail Rejected Already Voted',
    'C5m':	'By-mail Rejected No Voter ID',
    'C5n':	'By-mail Rejected No Ballot App',
    'State': 'State_Abbr'
}

# Keeping only the specified columns and renaming them
new_2020 = data_2020[columns_2020].rename(columns = columns_to_rename_2020)
new_2016 = data_2016[columns_2016].rename(columns = columns_to_rename_2016)

new_2020.to_csv('Cleaned_2020EAVSData.csv', index=False)
new_2016.to_csv('Cleaned_2016EAVSData.csv', index=False)

print("Datasets cleaned and saved successfully!")


Datasets cleaned and saved successfully!


### **Replacing -888888 in the 2016 dataset and -88 in the 2020 dataset with NA**

In [32]:
cleaned_2020 = pd.read_csv('Cleaned_2020EAVSData.csv')
cleaned_2016 = pd.read_csv('Cleaned_2016EAVSData.csv')

cleaned_2020.replace(-88, pd.NA, inplace=True)
cleaned_2016.replace('-888888: Not Applicable', pd.NA, inplace=True)

cleaned_2020.to_csv('Updated_Cleaned_2020EAVSData.csv', index=False)
cleaned_2016.to_csv('Updated_Cleaned_2016EAVSData.csv', index=False)

print("Special codes replaced with NA successfully!")

Special codes replaced with NA successfully!


### **Replacing -999999 in the 2016 dataset and -99 in the 2020 dataset with No Data**

In [33]:
cleaned_2020 = pd.read_csv('Updated_Cleaned_2020EAVSData.csv')
cleaned_2016 = pd.read_csv('Updated_Cleaned_2016EAVSData.csv')

cleaned_2020.replace(-99, 'No Data', inplace=True)
cleaned_2016.replace("-999999: Data Not Available", "No Data", inplace=True)

cleaned_2020.to_csv('Final_Cleaned_2020EAVSData.csv', index=False)
cleaned_2016.to_csv('Final_Cleaned_2016EAVSData.csv', index=False)

print("Special codes replaced with 'No Data' successfully!")


Special codes replaced with 'No Data' successfully!


### **Preview of Datasets**

In [34]:
cleaned_2020 = pd.read_csv('Final_Cleaned_2020EAVSData.csv')
cleaned_2016 = pd.read_csv('Final_Cleaned_2016EAVSData.csv')

# Displaying the first few rows to check cleaning results
print("Preview of 2020 Dataset:")
print(cleaned_2020.head())

print("\nPreview of 2016 Dataset:")
print(cleaned_2016.head())

Preview of 2020 Dataset:
    FIPSCode Jurisdiction_Name State_Full State_Abbr  \
0  100100000    AUTAUGA COUNTY    ALABAMA         AL   
1  100300000    BALDWIN COUNTY    ALABAMA         AL   
2  100500000    BARBOUR COUNTY    ALABAMA         AL   
3  100700000       BIBB COUNTY    ALABAMA         AL   
4  100900000     BLOUNT COUNTY    ALABAMA         AL   

  By-mail Transmitted Total By-mail Returned By Voters Total  \
0                    1329.0                           1026.0   
1                   11147.0                           9315.0   
2                     726.0                            647.0   
3                     332.0                            274.0   
4                    1032.0                            828.0   

  By-mail Returned Undeliverable By-mail Voided By-mail Voted In-Person  \
0                            0.0        No Data                    26.0   
1                           11.0        No Data                     0.0   
2                        No 

## Google Cloud Upload


In [35]:
# Grab Dependencies

from google.cloud import bigquery
from google.colab import auth
import pandas_gbq

# Initialize a BigQuery client
client = bigquery.Client(project='DS2002Final')

# auth.authenticate_user()
print('Authenticated')

# Upload to Google Big Query

pandas_gbq.to_gbq(cleaned_2020,'DS2002Final.2020EAVS', project_id='ds2002final-443502', if_exists='replace')
pandas_gbq.to_gbq(cleaned_2016,'DS2002Final.2016EAVS', project_id='ds2002final-443502', if_exists='replace')

Authenticated


100%|██████████| 1/1 [00:00<00:00, 8192.00it/s]


# ETL Setup and Implementation

### **Storing dataset in MongoDB**


In [36]:
cleaned_2020 = pd.read_csv('Final_Cleaned_2020EAVSData.csv')
cleaned_2016 = pd.read_csv('Final_Cleaned_2016EAVSData.csv')

connection_string = ######################################
client = MongoClient(connection_string)

db = client['eavs_db']

# Inserting data into collections
db.cleaned_2020.insert_many(cleaned_2020.to_dict('records'))
db.cleaned_2016.insert_many(cleaned_2016.to_dict('records'))

print("Data successfully loaded into MongoDB Atlas!")


Data successfully loaded into MongoDB Atlas!


### **Checking for Duplicates**


In [37]:
cleaned_2020 = pd.read_csv('Final_Cleaned_2020EAVSData.csv')
cleaned_2016 = pd.read_csv('Final_Cleaned_2016EAVSData.csv')

# Function to check for duplicates and insert only new data
def insert_if_not_exists(collection_name, dataframe, unique_field):
    collection = db[collection_name]

    for record in dataframe.to_dict('records'):
        # Check if the record already exists
        if not collection.find_one({unique_field: record[unique_field]}):
            collection.insert_one(record)

    print(f"Finished processing {collection_name}.")

insert_if_not_exists('cleaned_2020', cleaned_2020, 'FIPSCode')
insert_if_not_exists('cleaned_2016', cleaned_2016, 'FIPSCode')

print("Data successfully checked and inserted into MongoDB!")


Finished processing cleaned_2020.
Finished processing cleaned_2016.
Data successfully checked and inserted into MongoDB!


# Data Visualization

### EDA

In [38]:
import plotly.graph_objects as go
import plotly.express as px

#cleaned_2020 = pd.read_csv('Final_Cleaned_2020EAVSData.csv')
#cleaned_2016 = pd.read_csv('Final_Cleaned_2016EAVSData.csv')

cleaned_2020 = pandas_gbq.read_gbq("SELECT * FROM `ds2002final-443502.DS2002Final.2020EAVS`", project_id="ds2002final-443502")
cleaned_2016 = pandas_gbq.read_gbq("SELECT * FROM `ds2002final-443502.DS2002Final.2016EAVS`", project_id="ds2002final-443502")

cols_to_numeric_2020 = ['By-mail Transmitted Total', 'By-mail Returned By Voters Total', 'By-mail Counted Total', 'By-mail Rejected Total']

for i in cols_to_numeric_2020:
    cleaned_2020[i] = pd.to_numeric(cleaned_2020[i], errors='coerce')

bystate_2020 = cleaned_2020.groupby(['State_Abbr']).sum().reset_index()[['State_Abbr', 'By-mail Transmitted Total', 'By-mail Returned By Voters Total', 'By-mail Counted Total', 'By-mail Rejected Total']]

bystate_2020 = bystate_2020.drop([3, 8, 12, 27, 42, 50]).reset_index(drop = True).assign(Year = 2020) # Drops territories, adds year column

cols_to_numeric_2016 = ['By-mail Transmitted Total', 'By-mail Returned By Voters Total', 'By-mail Counted Total', 'By-mail Rejected Total']

for i in cols_to_numeric_2016:
    cleaned_2016[i] = pd.to_numeric(cleaned_2016[i], errors='coerce')

bystate_2016 = cleaned_2016.groupby(['State_Abbr']).sum().reset_index()[['State_Abbr', 'By-mail Transmitted Total', 'By-mail Returned By Voters Total', 'By-mail Counted Total', 'By-mail Rejected Total']]

bystate_2016 = bystate_2016.drop([7, 11, 40, 48]).reset_index(drop = True).assign(Year = 2016) # Drops territories, adds year column

bystate = pd.concat([bystate_2016, bystate_2020])

Downloading: 100%|██████████|
Downloading: 100%|██████████|


In [39]:
fig = go.Figure(data=go.Choropleth(
    locations=bystate_2020['State_Abbr'], # Spatial coordinates
    z = bystate_2020['By-mail Returned By Voters Total'].astype(float), # Data to be color-coded
    locationmode = 'USA-states', # set of locations match entries in `locations`
    colorscale = 'Reds',
    colorbar_title = "Votes"
))

fig.update_layout(
    title_text = 'Mail-In Ballots Submitted by Voters in 2020',
    geo_scope='usa', # limite map scope to USA
)

fig.show()

In [40]:
fig = go.Figure(data=go.Choropleth(
    locations=bystate_2016['State_Abbr'], # Spatial coordinates
    z = bystate_2016['By-mail Returned By Voters Total'].astype(float), # Data to be color-coded
    locationmode = 'USA-states', # set of locations match entries in `locations`
    colorscale = 'Blues',
    colorbar_title = "Votes"
))

fig.update_layout(
    title_text = 'Mail-In Ballots Submitted by Voters in 2016',
    geo_scope='usa', # limite map scope to USA
)

fig.show()

In [41]:
bystate['Percent of By-mail ballots accepted'] = 100*(1 - bystate['By-mail Rejected Total']/bystate['By-mail Returned By Voters Total'])
bystate['Percent of By-mail ballots rejected'] = 100*bystate['By-mail Rejected Total']/bystate['By-mail Returned By Voters Total']
bystate['Percent of By-mail ballots returned'] = 100*bystate['By-mail Returned By Voters Total']/bystate['By-mail Transmitted Total']
bystate

State_Abbr  By-mail Transmitted Total  By-mail Returned By Voters Total  \
0          AK                    31817.0                           27626.0   
1          AL                    98474.0                           88601.0   
2          AR                    29902.0                           27525.0   
3          AZ                  2478063.0                         2017722.0   
4          CA                 12018267.0                         8511992.0   
..        ...                        ...                               ...   
45         VT                   313193.0                          273784.0   
46         WA                  5042956.0                         4082581.0   
47         WI                  1441825.0                         1305082.0   
48         WV                   150202.0                          142445.0   
49         WY                    89540.0                           85627.0   

    By-mail Counted Total  By-mail Rejected Total  Year  \
0                 26750.0                   876.0  2016   
1                 87553.0                     0.0  2016   
2                 26655.0                  1614.0  2016   
3               1991683.0                 10769.0  2016   
4               8453683.0                 58309.0  2016   
..                    ...                     ...   ...   
45               272318.0                  1465.0  2020   
46              4051164.0                 31417.0  2020   
47              1302101.0                  2981.0  2020   
48               142191.0                   254.0  2020   
49                85454.0                   173.0  2020   

    Percent of By-mail ballots accepted  Percent of By-mail ballots rejected  \
0                             96.829074                             3.170926   
1                            100.000000                             0.000000   
2                             94.136240                             5.863760   
3                             99.466279                             0.533721   
4                             99.314978                             0.685022   
..                                  ...                                  ...   
45                            99.464907                             0.535093   
46                            99.230462                             0.769538   
47                            99.771585                             0.228415   
48                            99.821686                             0.178314   
49                            99.797961                             0.202039   

    Percent of By-mail ballots returned  
0                             86.827796  
1                             89.974003  
2                             92.050699  
3                             81.423354  
4                             70.825453  
..                                  ...  
45                            87.417024  
46                            80.956110  
47                            90.515978  
48                            94.835621  
49                            95.629886  

[100 rows x 9 columns]

###Visualizations

In [42]:
fig = go.Figure(data=go.Choropleth(
    locations = bystate[bystate['Year'] == 2016]['State_Abbr'], # Spatial coordinates
    z = bystate[bystate['Year'] == 2016]['Percent of By-mail ballots rejected'], # Data to be color-coded
    locationmode = 'USA-states', # set of locations match entries in `locations`
    colorscale = 'Greens',
    colorbar_title = 'Percent rejected'
))

fig.update_layout(
    title_text = 'Percent of Mail-In Ballots Rejected in 2016',
    geo_scope='usa', # limite map scope to USA
)

fig.show()

In [43]:
fig = go.Figure(data=go.Choropleth(
    locations = bystate[bystate['Year'] == 2020]['State_Abbr'], # Spatial coordinates
    z = bystate[bystate['Year'] == 2020]['Percent of By-mail ballots rejected'], # Data to be color-coded
    locationmode = 'USA-states', # set of locations match entries in `locations`
    colorscale = 'Purples',
    colorbar_title = 'Percent rejected'
))

fig.update_layout(
    title_text = 'Percent of Mail-In Ballots Rejected in 2020',
    geo_scope='usa', # limite map scope to USA
)

fig.show()

In [44]:
rejection_reasons_2016 = cleaned_2016

for i in rejection_reasons_2016.columns:
    rejection_reasons_2016[i] = pd.to_numeric(rejection_reasons_2016[i], errors='coerce')

rejection_reasons_2016 = rejection_reasons_2016.sum()[12:].reset_index().rename(columns = {0: 2016})

rejection_reasons_2020 = cleaned_2020

for i in rejection_reasons_2020.columns:
    rejection_reasons_2020[i] = pd.to_numeric(rejection_reasons_2020[i], errors='coerce')

rejection_reasons_2020 = rejection_reasons_2020.sum()[13:].reset_index().rename(columns = {'index': 'index1', 0: 2020})

rejection_reasons = pd.concat([rejection_reasons_2016, rejection_reasons_2020], axis = 1).drop(columns = 'index1')
rejection_reasons

index     2016      2020
0                 By-mail Rejected Deadline  73565.0   66957.0
1          By-mail Rejected Voter Signature  63897.0   66799.0
2        By-mail Rejected Witness Signature   9700.0   18378.0
3   By-mail Rejected Non-matching Signature  87647.0  157477.0
4          By-mail Rejected No EO Signature     60.0     132.0
5      By-mail Rejected Unofficial Envelope   2087.0   17715.0
6           By-mail Rejected Ballot Missing   2142.0    7240.0
7      By-mail Rejected Envelope Not Sealed    843.0    3109.0
8               By-mail Rejected No Address    898.0    2169.0
9         By-mail Rejected Multiple Ballots    222.0    8257.0
10                By-mail Rejected Deceased   4732.0    6599.0
11           By-mail Rejected Already Voted   4006.0   49016.0
12             By-mail Rejected No Voter ID   3460.0    7562.0
13           By-mail Rejected No Ballot App    250.0    1924.0

In [45]:
fig = px.bar(rejection_reasons,
             x="index",
             y= [2016, 2020],
             barmode='group')

fig.update_layout(
    title_text = 'Reasons for Mail-in Ballot Rejection - 2016 and 2020'
)

fig.show()

In [46]:
from plotly.subplots import make_subplots

fig = make_subplots(1, 2, specs=[[{'type':'domain'}, {'type':'domain'}]],
                    subplot_titles=['2016', '2020'])
fig.add_trace(go.Pie(labels = rejection_reasons['index'], values = rejection_reasons[2016]), 1, 1)
fig.add_trace(go.Pie(labels = rejection_reasons['index'], values = rejection_reasons[2020]), 1, 2)

fig.update_layout(title_text='Mail-In Ballot Rejection Reasons - 2016 and 2020')
fig.show()


### Statistical Analysis

Question: Are the rates of mail-in ballot rejection by state different between 2016 and 2020?

To answer this, we run a one-way ANOVA test with year as our independent variable.

In [47]:
import numpy as np
from scipy.stats import f_oneway

In [48]:
bystate

State_Abbr  By-mail Transmitted Total  By-mail Returned By Voters Total  \
0          AK                    31817.0                           27626.0   
1          AL                    98474.0                           88601.0   
2          AR                    29902.0                           27525.0   
3          AZ                  2478063.0                         2017722.0   
4          CA                 12018267.0                         8511992.0   
..        ...                        ...                               ...   
45         VT                   313193.0                          273784.0   
46         WA                  5042956.0                         4082581.0   
47         WI                  1441825.0                         1305082.0   
48         WV                   150202.0                          142445.0   
49         WY                    89540.0                           85627.0   

    By-mail Counted Total  By-mail Rejected Total  Year  \
0                 26750.0                   876.0  2016   
1                 87553.0                     0.0  2016   
2                 26655.0                  1614.0  2016   
3               1991683.0                 10769.0  2016   
4               8453683.0                 58309.0  2016   
..                    ...                     ...   ...   
45               272318.0                  1465.0  2020   
46              4051164.0                 31417.0  2020   
47              1302101.0                  2981.0  2020   
48               142191.0                   254.0  2020   
49                85454.0                   173.0  2020   

    Percent of By-mail ballots accepted  Percent of By-mail ballots rejected  \
0                             96.829074                             3.170926   
1                            100.000000                             0.000000   
2                             94.136240                             5.863760   
3                             99.466279                             0.533721   
4                             99.314978                             0.685022   
..                                  ...                                  ...   
45                            99.464907                             0.535093   
46                            99.230462                             0.769538   
47                            99.771585                             0.228415   
48                            99.821686                             0.178314   
49                            99.797961                             0.202039   

    Percent of By-mail ballots returned  
0                             86.827796  
1                             89.974003  
2                             92.050699  
3                             81.423354  
4                             70.825453  
..                                  ...  
45                            87.417024  
46                            80.956110  
47                            90.515978  
48                            94.835621  
49                            95.629886  

[100 rows x 9 columns]

In [49]:
#Note: We have to add dropna() since Alabama does not report the number of By-mail ballots rejected. n=49 for this test.

f_oneway(bystate['Percent of By-mail ballots rejected'][bystate['Year'] == 2016].dropna(), bystate['Percent of By-mail ballots rejected'][bystate['Year'] == 2020].dropna())

F_onewayResult(statistic=6.9028224213670075, pvalue=0.01000398684652135)

We got a p-value of 0.0100, which is low enough to indicate that the distributions changed. But how did they change?

In [50]:
fig = px.box(bystate, y='Percent of By-mail ballots rejected', x='Year', points = 'all')
fig.show()

From this, we see that though the highs and lows of the two distributions are similar, 2020 saw a large reduction in the percentage of mail-in ballots rejected across the states.

In [51]:
from google.cloud import bigquery
from google.colab import auth
import csv